src/algebra/monoid_algebra/division.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2022 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import algebra.monoid_algebra.basic
+import data.finsupp.order
+
+/-!
+# Division of `add_monoid_algebra` by monomials
+
+## Main definitions
+
+* `add_monoid_algebra.div_of g x`: divides `x` by the monomial `add_monoid_algebra.of k G g`
+* `add_monoid_algebra.mod_of g x`: the remainder upon dividing `x` by the monomial
+  `add_monoid_algebra.of k G g`.
+
+## Main results
+
+* `add_monoid_algebra.div_of_add_mod_of`, `add_monoid_algebra.mod_of_add_div_of`: `div_of` and
+  `mod_of` are well-behaved as quotient and remainder operators.
+
+## Implementation notes
+
+This file is most important for when `G = ℕ` (polynomials) or `G = σ →₀ ℕ` (multivariate
+polynomials). So as to generalize to both, we require
+```lean
+[canonically_ordered_add_monoid G] [has_sub G] [has_ordered_sub G] [contravariant_class G G (+) (≤)]
+```
+
+-/
+
+
+variables {k G : Type*} [semiring k]
+
+namespace add_monoid_algebra
+
+section
+variables [canonically_ordered_add_monoid G] [has_sub G] [has_ordered_sub G]
+  [contravariant_class G G (+) (≤)]
```

Comment:
```suggestion
variables [add_cancel_comm_monoid G] [has_le G] [contravariant_class G G (+) (≤)]
```
Reducing the assumptions to this allows you to remove several of the `letI` and works until you start needing `zero_le, preorder, exists_add_of_le`.

Reply:
I originally did this, but it seemed pointless; IMO division is meaningless without an associated notion of `mod` satisfying `mod_of g x + of' k G g * div_of g x`, so better to use the typeclasses needed for that statement throughout.

Reply:
One of my use cases is Laurent polynomials, so I try to keep open the possibility of groups for as long as possible.